File: cpp/515.FindLargestValueinEachTreeRow.cpp
#include <algorithm>
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        
        vector<int> ret; // 创建就初始化，又不是指针，这一点跟java很不一样
        
        dfs(root, ret, 0);
        
        return ret;
    }
   // &ret 引用传递，java默认就是传递引用 
    void dfs(TreeNode* node, vector<int> &ret, int level){
        if (node == NULL) return; // c++的NULL，对应java 的null，swift的nil，python的None
        
        int val = node->val;
        if (ret.size() == level){
            ret.push_back(val);
        }else{
            ret[level] = max(val, ret[level]);
        }
        
        dfs(node->left, ret, level+1);
        dfs(node->right, ret, level+1);
    }
};

File: js/515.FindLargestValueinEachTreeRow.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 // DFS 139 ms
var largestValues = function(root) {
    var maxes = []
    dfs(root, 0, maxes)
    return maxes
};

var dfs = function(root, level, maxes) {
    if (!root) return
    
    var val = root.val
    if (maxes.length == level){
        maxes.push(val)    // swift: append, python: append, js:push, java: add, c++:push_back
    }else{
        maxes[level] = Math.max(val, maxes[level]) // 用Math 的函数，得Math.xxx()
    }
    
    dfs(root.left, level+1, maxes)
    dfs(root.right, level+1, maxes)
}